File: 2018-2019/zima/PF/lista2/zad6.ml
let rec suffixes xs =
    match xs with
    | [] -> [[]]
    | x::xs -> (x::xs) :: suffixes xs

let rec map f xs =
    match xs with
    | []    -> []
    | y::ys -> f y :: map f ys

let rec prefixes xs =
    match xs with
    | []    -> [[]]
    | y::ys -> [] :: map (fun zs -> y::zs) (prefixes ys)

(* let rec prefixes xs =
    let rec reverse ?(acc = []) xs =
        match xs with
        | []    -> acc
        | x::xs -> reverse ~acc:(x :: acc) xs
    in let rec map f xs =
        match xs with
        | []    -> []
        | x::xs -> (f x) :: map f xs
    in reverse (map reverse (suffixes (reverse xs))) *)

File: 2018-2019/lato/Prolog/L2/zad6.pl
reverse(Xs, Ys) :- 
    reverse(Xs, [], Ys, Ys).

reverse([], Ys, Ys, []).
reverse([X | Xs], Acc, Ys, [_ | YsGuard]) :-
    reverse(Xs, [X | Acc], Ys, YsGuard).
